aspnetcore/mvc/models/file-uploads.md

```diff
@@ -1,195 +1,654 @@
+</form>
+```
 
-Copy the `IFormFile` to a stream and save it to the byte array:
+When the form is POSTed to the server, copy the <xref:Microsoft.AspNetCore.Http.IFormFile> to a stream and save it as a byte array in the database:
 
 ```csharp
-// POST: /Account/Register
-[HttpPost]
-[AllowAnonymous]
-[ValidateAntiForgeryToken]
-public async Task<IActionResult> Register(RegisterViewModel model)
+public async Task<IActionResult> OnPostUploadAsync()
 {
-    ViewData["ReturnUrl"] = returnUrl;
-    if  (ModelState.IsValid)
+    using (var memoryStream = new MemoryStream())
     {
-        var user = new ApplicationUser 
+        await FileUpload.FormFile.CopyToAsync(memoryStream);
+    
+        var file = new AppFile()
         {
-            UserName = model.Email,
-            Email = model.Email
+            Content = memoryStream.ToArray()
         };
-        using (var memoryStream = new MemoryStream())
-        {
-            await model.AvatarImage.CopyToAsync(memoryStream);
-            user.AvatarImage = memoryStream.ToArray();
-        }
-    // additional logic omitted
-
-    // Don't rely on or trust the model.AvatarImage.FileName property 
-    // without validation.
+        
+        _context.File.Add(file);
+        await _context.SaveChangesAsync();
+    }
 }
 ```
 
-> [!NOTE]
+> [!WARNING]
 > Use caution when storing binary data in relational databases, as it can adversely impact performance.
+>
+> Don't rely on or trust the `FileName` property of <xref:Microsoft.AspNetCore.Http.IFormFile> without validation. The `FileName` property should only be used for display purposes and only after HTML encoding.
+>
+> The examples provided don't take into account security considerations. See the [Security considerations](#security-considerations) and [Validation](#validation) sections and the sample app for information and examples of file upload validation.
 
-## Uploading large files with streaming
+### Upload large files with streaming
 
-If the size or frequency of file uploads is causing resource problems for the app, consider streaming the file upload rather than buffering it in its entirety, as the model binding approach shown above does. While using `IFormFile` and model binding is a much simpler solution, streaming requires a number of steps to implement properly.
+The following example demonstrates using JavaScript to stream a file to a controller action. The file's antiforgery token is generated using a custom filter attribute and passed to the client HTTP headers instead of in the request body. Because the action method processes the uploaded data directly, model binding is disabled by another custom filter. Within the action, the form's contents are read using a `MultipartReader`, which reads each individual `MultipartSection`s, processing the file or storing the contents as appropriate. Once all of the sections are read, the action performs its own model binding.
```

Comment:
The filter only disables form model binding not all binding. It is still possible to info passed in the url to parameters.
```suggestion
The following example demonstrates using JavaScript to stream a file to a controller action. The file's antiforgery token is generated using a custom filter attribute and passed to the client HTTP headers instead of in the request body. Because the action method processes the uploaded data directly, form model binding is disabled by another custom filter. Within the action, the form's contents are read using a `MultipartReader`, which reads each individual `MultipartSection`s, processing the file or storing the contents as appropriate. Once all of the sections are read, the action performs its own model binding.
```